### PR TITLE
Restore atoi64/GetIntArg saturating behavior

### PR DESCRIPTION
**Pull request by jamesob**
The new locale-independent atoi64 method introduced in #20452 behaves
differently for values passed which are greater than the uint64_t max.
This commit is proof of that (the attached test fails), and is meant to spur discussion 
on how to handle such an incompatibility.

The change as committed in #20542 may break some end-usages of
bitcoind, but more than that may have consensus implications as this
deserialization mechanism is used in CScript::ParseScript. ~~I think this
means it's possible that #20542 could have been an accidental soft fork?~~

Edit: `CScript::ParseScript` isn't consensus-critical, but is used in e.g. script descriptor parsing.

### CODE REVIEW

**Comment by sipa** [comment]
Nice catch.

CScript::ParseScript shouldn't be consensus-critical though.

**Comment by jamesob** [comment]
Some discussion from IRC:
```
12:21 <MarcoFalke> jamesob: Integer overflow inside atoi is UB or at least unspecified behavior, last time I checked
12:22 <MarcoFalke> That is: Parsing 999999999999 can give you any value (it doesn't have to be 0 or max or 999999999999 mod max)
12:23 <MarcoFalke> jamesob: It is also not used in consensus code, so this is not a consensus change
12:24 <jamesob> this is not consensus? https://github.com/bitcoin/bitcoin/pull/20452/files#diff-1db27ed1bfbf61ea0fe64447413ef9f24238be710e7ca4ae9c7bc7a5c994eca0L72-R72
12:26 <jamesob> so here's the tricky thing: our atoi64 wasn't actually using atoi, but strtoll - for which oveflow behavior is well defined: "The strtol() function returns the result of the conversion, unless the value would underflow or overflow. If an underflow occurs, strtol() returns LONG_MIN. If an overflow occurs, strtol() returns LONG_MAX."
12:26 <jamesob> ("The strtoll() function works just like the strtol() function but returns a long long integer value.")
12:28 <MarcoFalke> Yeah, not consensus. Could be used by bitcoin-util or some RPC function.
12:29 <jamesob> Phew
```

**Comment by MarcoFalke** [comment]
Indeed good catch, I wasn't aware it is safe and saturating to produce an overflow. Still, an explicit error would be better than silent saturation. See also https://github.com/bitcoin/bitcoin/pull/17385/files#diff-19427b0dd1a791adc728c82e88f267751ba4f1c751e19262cac03cccd2822216R509

**Comment by jamesob** [comment]
I've pushed a change that restores compatibility by introducing a new function, `LocaleIndependentAtoi64`. This is specifically not a template specialization of its more generic cousin because the under-/overflow behavior differs (as was the case previously with our `atoi64` implementation). A compile error is rendered if someone attempts to use the generic version with `int64_t` types.

I've also added tests for `GetIntArg` that assert the saturation behavior.

When this is merged I'll make a follow-up issue to surface InitErrors instead of saturating, which would be a good first issue I think.

**Comment by MarcoFalke** [comment]
If you want this merged, please make sure that each commit compiles and passes all tests. Otherwise it will break git bisect.

**Review by maflcko — COMMENTED** [review]
I think you can squash the remaining test commit into the feature commit. No need for a fresh commit.

Also, left a question.

(Note to myself: commit 5b78055d5965b6ed69c6160100b1e4e36e6e6ea3 was removed in the latest rebase)

File: src/test/getarg_tests.cpp [hunk 1]
```diff
@@ -140,9 +141,12 @@ BOOST_AUTO_TEST_CASE(intarg)
     BOOST_CHECK_EQUAL(m_local_args.GetIntArg("-foo", 11), 11);
     BOOST_CHECK_EQUAL(m_local_args.GetIntArg("-foo", 0), 0);
 
-    ResetArgs("-foo -bar");
-    BOOST_CHECK_EQUAL(m_local_args.GetIntArg("-foo", 11), 0);
-    BOOST_CHECK_EQUAL(m_local_args.GetIntArg("-bar", 11), 0);
+    BOOST_CHECK_EQUAL(m_local_args.GetIntArg("-foo", 0), 0);
```
[the remark this answers is not in the document]
Comment (jamesob) [reply src/test/getarg_tests.cpp:144]: Oops, this looks like a mistaken edit.

**Comment by laanwj** [comment]
Concept NACK on reintroducing `strtoll` usage. I would prefer implementing the desired compatible behavior in an explicit way using new functions, instead of delegating it to the C library's potentially locale-dependent function.

File: src/test/getarg_tests.cpp [hunk 2]
```diff
@@ -144,6 +145,11 @@ BOOST_AUTO_TEST_CASE(intarg)
     BOOST_CHECK_EQUAL(m_local_args.GetIntArg("-foo", 11), 0);
     BOOST_CHECK_EQUAL(m_local_args.GetIntArg("-bar", 11), 0);
 
+    // Check under-/overflow behavior.
+    ResetArgs("-foo=-9223372036854775809 -bar=9223372036854775808");
```
Comment (MarcoFalke) [inline src/test/getarg_tests.cpp:149]: This is not overflowing. It is equal to the min/max?

**Comment by jamesob** [comment]
@laanwj I think there is some confusion here: nowhere in the live code do I actually use `strtoll`; it's just used in tests to model compatibility with the old implementation in previous versions.

File: src/test/getarg_tests.cpp (hunk 2 above)
Comment (jamesob) [reply src/test/getarg_tests.cpp:149]:
```cpp
10:27:37 james@slug james/tmp % cat int64max.cpp
#include <limits>
#include <iostream>
#include <cinttypes>

int main() {
    std::cout << std::numeric_limits<int64_t>::max() << '\n';
}
10:27:44 james@slug james/tmp % ./a.out
9223372036854775807
```

**Review by maflcko — APPROVED** [review]
ack

File: src/util/strencodings.cpp [hunk 3]
```diff
@@ -565,3 +565,31 @@ std::optional<uint64_t> ParseByteUnits(const std::string& str, ByteUnit default_
...
+    // Emulate atoi(...) handling of white space and leading +/-.
+    std::string s = TrimString(str);
+    if (!s.empty() && s[0] == '+') {
+        if (s.length() >= 2 && s[1] == '-') {
+            return 0;
+        }
+        s = s.substr(1);
+    }
+
+    auto [_, error_condition] = std::from_chars(s.data(), s.data() + s.size(), result);
```
Comment (JeremyRubin) [inline src/util/strencodings.cpp:581]: std::from_chars uses strtol locale dependent behavior, no?

Comment (JeremyRubin) [reply src/util/strencodings.cpp:581]: https://en.cppreference.com/w/cpp/utility/from_chars says

> 1) Integer parsers: Expects the pattern identical to the one used by std::strtol in the default ("C") locale and the given non-zero numeric base, except that

Comment (sipa) [reply src/util/strencodings.cpp:581]: It always uses the "C" locale, so it is locale-independent.

Comment (JeremyRubin) [reply src/util/strencodings.cpp:581]: gotcha.

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/23841/commits/ac1a5b113ccdace4d7f04e2e8c4089e397b5c0b4

**Comment by MarcoFalke** [comment]
Edited title

**Comment by MarcoFalke** [comment]
@laanwj Are you still NACK on this? Otherwise I think this is rfm

**Review by shaavan — APPROVED** [review]
Code Review ACK ac1a5b113ccdace4d7f04e2e8c4089e397b5c0b4

According to Microsoft [docs](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/atoi64-atoi64-l-wtoi64-wtoi64-l?view=msvc-170):

> In the case of overflow with large positive integral values, **_atoi64** returns **I64_MAX** and **I64_MIN** in the case of overflow with large negative, integral values.

However, this was not true for the **LocaleIndependentAtoi<int64_t>** function, which intended to replicate **_atoi64** behavior. This function returns **0** both in case of underflow and overflow.

This PR fixes it by introducing a slightly modified **LocaleIndependentAtoi** function specifically for **int64_t** which correctly replicates **_atoi64** behavior of return **INT64_MAX** in case of overflow, and **INT64_MIN** in case of the underflow.

The added tests make sense and work perfectly, which, as expected, failed on the Master branch.
Thanks for catching this and preventing a potentially major future issue.

File: src/util/strencodings.h [hunk 4]
```diff
@@ -115,6 +118,21 @@ T LocaleIndependentAtoi(const std::string& str)
...
+ * template-specialization of the above function.
+ *
+ * New code should use ToIntegral or the ParseInt* functions which provide
+ * parse error feedback.
+ *
+ * This aims to replicate the exact behaviour of atoi64 in previous versions of
+ * Bitcoin Core. The old atoi64 utility function actually made use of `strtoll`
+ * in its implementation (or Microsoft's `_atoi64`), which saturates over- and
+ * underflows. That behavior is reproduced here.
+ */
```
Comment (ryanofsky) [inline src/util/strencodings.h:133]: In commit "Restore atoi64 compatibility with old versions of Bitcoin Core" (ac1a5b113ccdace4d7f04e2e8c4089e397b5c0b4)

I find this description hard to follow. What is the point of referencing all these other functions and vaguely handwaving about what this function does of simply saying "Parse a decimal integer string. On positive overflow, return the maximum integer value, on negative overflow return the minimum integer value, and on error, return 0."?

File: src/util/strencodings.h [hunk 5]
```diff
@@ -94,8 +94,11 @@ void SplitHostPort(std::string in, uint16_t& portOut, std::string& hostOut);
 // which provide parse error feedback.
 //
 // The goal of LocaleIndependentAtoi is to replicate the exact defined behaviour
-// of atoi and atoi64 as they behave under the "C" locale.
-template <typename T>
+// of atoi as it behaves under the "C" locale.
+//
+// Prevent use of LocaleIndependentAtoi for int64_t, since the function below
```
Comment (ryanofsky) [inline src/util/strencodings.h:99]: In commit "Restore atoi64 compatibility with old versions of Bitcoin Core" (ac1a5b113ccdace4d7f04e2e8c4089e397b5c0b4)

I don't understand why this function exists at all. What use could there be for a parsing function that returns 0 in the case of overflow? It just seems like a footgun. Returning the biggest possible value when a bigger value is given, and the smallest possible value when a smaller value is given seems like a more useful and less dangerous way to parse values.

I guess it precedes this PR, but the naming also seems strange. Why call this one util function locale independent, when basically every util function we have should be locale independent? If we had use for a parsing function that did vary by locale it would make more sense to draw attention and mention locales in **that** function's name, not this function's name.

File: src/util/strencodings.h [hunk 6]
```diff
@@ -115,6 +118,21 @@ T LocaleIndependentAtoi(const std::string& str)
     return result;
 }
 
+/**
+ * LocaleIndependentAtoi64 is provided for backwards compatibility reasons.
+ * Since its behavior differs from standard atoi functionality, it is not a
```
Comment (ryanofsky) [inline src/util/strencodings.h:123]: In commit "Restore atoi64 compatibility with old versions of Bitcoin Core" (ac1a5b113ccdace4d7f04e2e8c4089e397b5c0b4)

I can't figure how this function is not a valid implementation of the standard function https://en.cppreference.com/w/cpp/string/byte/atoi. If its behavior is actually nonstandard, it would be good to say how.

**Review by ryanofsky — APPROVED** [review]
Concept ~0. I don't understand why this PR is forking the parse function and only removing the overflow footgun specifically in the `int64_t`  case and letting it remain in other cases, instead of parsing sanely in all cases. I also think the commit descriptions and function descriptions are handwavy and confusing. They should just say what actual behaviors are instead of gesturing vaguely at other functions and PRs without saying what the differences are.

Code review ACK ac1a5b113ccdace4d7f04e2e8c4089e397b5c0b4, though. This does seem like the smallest possible change that will fix a few int64 parses, and it has good test coverage.

**Comment by MarcoFalke** [comment]
> I don't understand why this PR is forking the parse function and only removing the overflow footgun specifically in the int64_t case

In the int case it previously used `atoi`, which runs into UB (not saturation). See https://github.com/bitcoin/bitcoin/pull/23841#issuecomment-999747601 . So I presume the goal was to make minimal changes?


Though, given that it is UB I can also see the argument to prefer less code and turn the UB into saturation. (In master it was turned to return `0` IIRC).

File: src/util/strencodings.h (hunk 5 above)
Comment (MarcoFalke) [reply src/util/strencodings.h:99]: I like that the name is long and ugly. It makes developers less likely to pick it, when there are better named and safer alternatives available that come with error reporting instead of implicitly mapping errors to random (in-range) return values.

Comment (ryanofsky) [reply src/util/strencodings.h:99]: re: https://github.com/bitcoin/bitcoin/pull/23841#discussion_r781924579

> I like that the name is long and ugly. It makes developers less likely to pick it, when there are better named and safer alternatives available that come with error reporting instead of implicitly mapping errors to random (in-range) return values.

It's just seems a weird way to make the function name long by adding a random fact about locale independence. You could call it `FrogsAreGreenAtoi` and that would be better name than `LocaleIndependentAtoi` because it wouldn't falsely suggest that other functions depend on the current locale and this function doesn't. I think an ideal name would be something more like `ParseIntUnchecked` or `AtoiIgnoreErrors`

**Review by ryanofsky — COMMENTED** [review]
re: https://github.com/bitcoin/bitcoin/pull/23841#issuecomment-1009736820

> Though, given that it is UB I can also see the argument to prefer less code and turn the UB into saturation. (In master it was turned to return `0` IIRC).

Yes, that's what I'm trying to figure out. I don't understand why you would return 0 when the provided value is bigger than the maximum value, instead of returning the maximum value.  Also why you would want two error-ignoring parsing functions with different and not clearly documented behavior instead of one error-ignoring parsing function with clearly documented behavior? And is there actually something different about the int64_t type that it needs to be treated differently than all the other integer types?

Anyway, I acked this PR since it seems OK to merge, but I think it would be nicer if the two functions were consolidated, given nicer behavior and documentation, and probably renamed.

**Comment by MarcoFalke** [comment]
>  is there actually something different about the int64_t type that it needs to be treated differently than all the other integer types?

No, not that I am aware of.

File: src/util/strencodings.h (hunk 5 above)
Comment (MarcoFalke) [reply src/util/strencodings.h:99]: > FrogsAreGreenAtoi

Ryan, not all frogs are green. I'd suggest to use the name `MyFrogIsGreenAtoi` or `ParseSaturatedInt`.

Comment (ryanofsky) [reply src/util/strencodings.h:99]: +1 to `MyFrogIsGreenAtoi`

**Review by maflcko — CHANGES_REQUESTED** [review]
Withdraw my ACK. Seems more minimalistic to have one method instead of two.

**Comment by jamesob** [comment]
I think there is some navel-gazing going on here. We are in a situation where a change has inadvertently broken compatibility with a prior release (albeit in a minor way). This change proposes to remedy that and is well-tested.

If there are style concerns, I think they can be done in a follow-up. Otherwise I have other things to do and someone else can fix this in a way that reviewers find preferable.

**Comment by MarcoFalke** [comment]
>  are you proposing to special case this single method on the basis of type?

No, simply make the existing method saturate instead of creating yet another one. See https://github.com/bitcoin/bitcoin/pull/23841#pullrequestreview-849309409

If not done here, there will be a follow-up doing it. And I think that we should avoid follow-ups that completely remove code that was added previously. (They are fine when only a subset is followed-up on).

If you don't have time or energy to fix it that is fine, just let us know so that it can be picked up.

**Comment by ryanofsky** [comment]
> I think there is some navel-gazing going on here. 

It's true, and to make the navel-gazing even worse, it was mostly about the already merged PR #20452, not even this PR!

Anyway posted #24041 as an alternative

**Comment by MarcoFalke** [comment]
The code here is non-trivial to review, based on the bug that was accidentally introduced.

So I don't think it is navel-gazing to consolidate the tricky code to just one function instead of two.

Also, I don't think it is navel-gazing to prefer a +63-17 patch over a two patches (each of which is larger than the single one), first +88-13, then +42-71.

**Comment by jamesob** [comment]
Closing in favor of https://github.com/bitcoin/bitcoin/pull/24041

**Comment by JeremyRubin** [comment]
if i understand it correctly, i do prefer the approach here as a stop-gap compared to #24041.

What I think makes the most sense is to do both of these (this PR first) and then over the course of a few releases:

1. Log Warn if the parsing will differ in a future release
2. Error + Shutdown if the parsing will differ in a future release
3. Change the parsing.

Perhaps it's more trouble than it's worth to DTRT here, but I think it is, personally, since config files breaking is a big deal.

**Comment by ryanofsky** [comment]
> Perhaps it's more trouble than it's worth to DTRT here, but I think it is, personally, since config files breaking is a big deal.

I don't think there are any differences between this PR and #24041 with respect to config files. Config behavior unintentionally changed in #20452, and both this PR and #24041 restore the previous behavior.

**Comment by JeremyRubin** [comment]
ok i was reading off of https://github.com/bitcoin/bitcoin/pull/24041#issuecomment-1010210872 but it seems that is not the case, in which case i'll concept ack #24041

**Comment by laanwj** [comment]
> @laanwj I think there is some confusion here: nowhere in the live code do I actually use strtoll; it's just used in tests to model compatibility with the old implementation in previous versions.

> @laanwj Are you still NACK on this? Otherwise I think this is rfm

No, I'm not, using it in the tests is fine sorry for misunderstanding it, my brain doesn't work anymore.
